### ISSUE TITLE
BUG : System.IdentityModel.Tokens.Jwt 6.12.1 causes exceptions

### ISSUE THREAD

**Issue by ErikPilsits-RJW** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#1686]
I have several API projects where this package was installed. I recently updated to 6.12.1 from 6.12.0 through nuget and started getting exceptions.
```
System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'System.String'.
 ---> System.MissingMethodException: Method not found: 'Void Microsoft.IdentityModel.Logging.IdentityModelTelemetryUtil.SetTelemetryData(System.Net.Http.HttpRequestMessage)'.
   at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
   --- End of inner exception stack trace ---
   at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.AuthenticateAsync()
   at Microsoft.AspNetCore.Authentication.AuthenticationService.AuthenticateAsync(HttpContext context, String scheme)
   at Microsoft.AspNetCore.Authorization.Policy.PolicyEvaluator.AuthenticateAsync(AuthorizationPolicy policy, HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
```

Oddly, I found that I actually no longer need this package at all, and have removed it, thus solving my issue. But I'm wondering if 6.12.1 has some dependency problem when installed.

**Comment by henrik-me** [comment]
@RojaEnnam

**Comment by RojaEnnam** [comment]
@ErikPilsits-RJW IdentityModelTelemetryUtil.SetTelemetryData is an internal method that we changed the method signature of in 6.12.1, can you please check if you have upgraded all the related nugets, specially Microsoft.IdentityModel.Logging and Microsoft.IdentityModel.Protocols

**Comment by BlacKCaT27** [comment]
Can confirm this is having a severe impact on our systems builds as well. https://www.nuget.org/packages/System.IdentityModel.Tokens.Jwt/6.12.1 shows that this new version is being served by NuGet but isn't downloading/working. NuGet tries to serve this version but it doesn't work so builds are breaking.

**Comment by ErikPilsits-RJW** [comment]
> @ErikPilsits-RJW IdentityModelTelemetryUtil.SetTelemetryData is an internal method that we changed the method signature of in 6.12.1, can you please check if you have upgraded all the related nugets, specially Microsoft.IdentityModel.Logging and Microsoft.IdentityModel.Protocols

We do not have those packages as separate dependencies. However this upgrade was done as a whole solution upgrade, so everything was upgraded at the same time.

**Comment by henrik-me** [comment]
6.12.1 packages unlisted from nuget.org. 
@BlacKCaT27  @ErikPilsits-RJW - do you have a repro? which packages are you referencing?

**Comment by henrik-me** [comment]
@BlacKCaT27 @ErikPilsits-RJW we are not able to reproduce this issue. Did you update all packages to the latest version? Can you please share a list of all your package references and versions to packages named Microsoft.IdentityModel and System.IdentityModel?

**Comment by BlacKCaT27** [comment]
In my case, it may have been related to a cached instance of our repository
manager on my local machine. Once the bad package was removed from nuget
and I invalidated my cache, builds succeeded again.

My repro was to reference a dependency which referenced this package.
Beyond that I can't say unfortunately.

**Comment by henrik-me** [comment]
Thanks @BlacKCaT27, that would match our expectations as well. If all packages are updated there should be no such issues. We will re-enlist later today unless we hear from you or others that the issue persists.

**Comment by BlacKCaT27** [comment]
Can confirm the issue is resolved once Nexus cache was cleared. Not sure
what went wrong with the package itself though. Good luck

**Comment by henrik-me** [comment]
Thanks, will close this one and ensure the packages are listed again.

**Comment by ErikPilsits-RJW** [comment]
Bad cache and builds was not my issue. Mine was a runtime exception. As I said, I updated all packages in my entire solution successfully, and encountered the runtime error. Reverting to version 6.12.0 resolved the runtime exception. Also, removing the package completely, as I said I found I didn't actually need it anymore, also resolved the issue.

Is it possible other dependent packages are failing due to the method signature change if they happen to take the dependency on this package but don't lock a version?

I will try to get a dependency list for you, but this package was a dependency in a core project and so the dependency graph is complex.

**Comment by ErikPilsits-RJW** [comment]
> Thanks, will close this one and ensure the packages are listed again.

Please don't close this, it's not fixed.

**Comment by henrik-me** [comment]
re-opened, have re-listed the packages.

**Comment by ErikPilsits-RJW** [comment]
Api web project dependencies, which depends on infrastructure and core projects
```
  <ItemGroup>
    <PackageReference Include="AutoMapper" Version="10.1.1" />
    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="8.1.1" />
    <PackageReference Include="FluentValidation.AspNetCore" Version="10.3.0" />
    <PackageReference Include="Hellang.Middleware.ProblemDetails" Version="6.1.0" />
    <PackageReference Include="LazyCache.AspNetCore" Version="2.1.3" />
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.18.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.18" />
    <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.8" />
    <PackageReference Include="Microsoft.Extensions.Http.Polly" Version="5.0.1" />
    <PackageReference Include="Microsoft.Identity.Web" Version="1.15.2" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="Polly" Version="7.2.2" />
    <PackageReference Include="Polly.Extensions.Http" Version="3.0.0" />
    <PackageReference Include="RJW.QueryBuilder" Version="0.2.6" />
    <PackageReference Include="Serilog.AspNetCore" Version="4.1.0" />
    <PackageReference Include="Serilog.Settings.Configuration" Version="3.2.0" />
    <PackageReference Include="Serilog.Sinks.ApplicationInsights" Version="3.1.0" />
    <PackageReference Include="Serilog.Sinks.Console" Version="4.0.0" />
    <PackageReference Include="Serilog.Sinks.File" Version="5.0.0" />
    <PackageReference Include="Serilog.Sinks.MSSqlServer" Version="5.6.0" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.1.5" />
    <PackageReference Include="Swashbuckle.AspNetCore.Annotations" Version="6.1.5" />
    <PackageReference Include="Swashbuckle.AspNetCore.Newtonsoft" Version="6.1.5" />
  </ItemGroup>
```

Infrastructure
```
  <ItemGroup>
    <PackageReference Include="Dapper" Version="2.0.90" />
    <PackageReference Include="LazyCache.AspNetCore" Version="2.1.3" />
    <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.8" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="3.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Http.Polly" Version="5.0.1" />
    <PackageReference Include="Microsoft.Extensions.Options" Version="5.0.0" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="Polly" Version="7.2.2" />
    <PackageReference Include="RJW.QueryBuilder" Version="0.2.6" />
  </ItemGroup>
```

Core
```
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Http.Features" Version="5.0.9" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="3.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="5.0.0" />
    <PackageReference Include="Microsoft.Extensions.Options" Version="5.0.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="Polly" Version="7.2.2" />
    <PackageReference Include="RJW.QueryBuilder" Version="0.2.6" />
    <PackageReference Include="System.Linq.Async" Version="5.0.0" />
  </ItemGroup>
```

QueryBuilder is an internal project on our enterprise nuget feed
```
  <ItemGroup>
    <PackageReference Include="FluentValidation" Version="10.1.0" />
    <PackageReference Include="JsonSubTypes" Version="1.8.0" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="3.1.14" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="3.1.14" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
  </ItemGroup>
```

The exception occurs if I add System.IdentityModel.Tokens.Jwt to the Infrastructure dependencies.

**Comment by henrik-me** [comment]
@ErikPilsits-RJW, which exact package did you remove to make things work?

**Comment by henrik-me** [comment]
```
 <PackageReference Include="Microsoft.Identity.Web" Version="1.15.2" /> 
``` 
references IdentityModel, current latest release references 6.12.0

**Comment by ErikPilsits-RJW** [comment]
I removed or downgraded to 6.12.0 this package _only_ , System.IdentityModel.Tokens.Jwt, in the Infrastructure project.

**Comment by henrik-me** [comment]
CC: @jennyf19 

Microsoft.Identity.Web, [references](https://github.com/AzureAD/microsoft-identity-web/blob/ce9d7322072b0ffd08cd8abd9027c958d1a52e1b/src/Microsoft.Identity.Web/Microsoft.Identity.Web.csproj)
```
    <PackageReference Include="Microsoft.IdentityModel.Logging" Version="6.12.0" />
    <PackageReference Include="Microsoft.IdentityModel.Protocols.OpenIdConnect" Version="6.12.0" />
    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.12.0" />
```

**Comment by henrik-me** [comment]
Thanks @ErikPilsits-RJW , I believe we know why it helped removing the direct reference.

**Comment by ErikPilsits-RJW** [comment]
Seems like you found it. Now that you relisted the package, I recreated my issue. I confirmed that after building, I only have version 6.12.1 of Tokens.Jwt.dll in my solution output. So the method sig change error makes sense.

Seems like this kind of a change is really dangerous no? It will break anything that has a dependency on this package and uses that method.

**Comment by henrik-me** [comment]
Agree it can be dangerous as can be seen here.  All our packages needs to be updated at the same time. 

Will explore if we can remove version limitation for Microsoft.Identity.Web. @jennyf19  ?

**Comment by henrik-me** [comment]
Keeping this open as we explore an option to provide better messages in such cases, avoiding the issue all together or something else.

**Comment by ErikPilsits-RJW** [comment]
This is a bit odd though. It's an internal method, Microsoft.Identity.Web cannot call it directly. So even a dependency on version 6.12.0 of this package cannot break it like that.

**Comment by henrik-me** [comment]
Agree, we have internal utility functions in one place. 
We should be able to avoid this issue, however it's also a potential issue if the packages are not kept in sync.

**Comment by ErikPilsits-RJW** [comment]
Now I understand. From Microsoft.IdentityModel.Logging
```
[assembly: InternalsVisibleTo("Microsoft.IdentityModel.Protocols
```
And my Protocols version is 6.12.0 from the Microsoft.Identity.Web dependency. And so the signatures don't match.

So really, this is a potential problem for the wider world. Consider two unrelated packages, and the maintainers are dead. One takes a dependency on Protocols 6.12.0, and one takes a dependency on Tokens.Jwt 6.12.1. You'd have to fix it by taking an explicit dependency on Protocols 6.12.1, and hope that Protocols doesn't have breaking changes for your other dependency.

Wouldn't it be safer to maintain backwards compatibility and provide both method signatures?

**Comment by ErikPilsits-RJW** [comment]
Confirmed. If I take another explicit dependency on
```
<PackageReference Include="Microsoft.IdentityModel.Protocols" Version="6.12.1" />
```
the exception is resolved.

**Comment by henrik-me** [comment]
> Wouldn't it be safer to maintain backwards compatibility and provide both method signatures?

Agree. We will look at how we can do. 
The mismatch of versions between packages could cause transparent issues and keeping the packages in sync is also important. One possibility is to keep the method signatures and in many cases that will be fine, there are situations we know versions must match in those cases we can throw much better exceptions.

**Comment by pkwieci1** [comment]
Can confirm the issue, I had to wrestle with the problem for the whole day until I found the culprit.
Microsoft.IdentityModel.Tokens works with version 6.12.0, but 6.12.1 causes exceptions breaking the app.

I have entirely deleted the package reference from the project and it's back working fine.

**Comment by nbiada** [comment]
I can confirm that the bug exists without a direct reference of the packages.
With this configuration in my _Infrastructure_ project it works. An entire day lost to find this. I hate you ;-))
```xml
  <ItemGroup>
    <PackageReference Include="AutoMapper" Version="10.1.1" />
    <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="8.1.1" />
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="5.0.9" />
    <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="2.2.0" />
    <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="5.0.9" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.9" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="5.0.9" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="5.0.9">
      <PrivateAssets>all</PrivateAssets>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Extensions.Identity.Stores" Version="5.0.9" />
    <PackageReference Include="Microsoft.Extensions.Localization.Abstractions" Version="5.0.9" />
    <PackageReference Include="Microsoft.Graph" Version="4.3.0" />
    <PackageReference Include="Microsoft.Graph.Auth" Version="1.0.0-preview.7" />
    <PackageReference Include="Microsoft.Identity.Client" Version="4.35.1" />
    <PackageReference Include="MimeKit" Version="2.15.0" />
    <PackageReference Include="MailKit" Version="2.15.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
    <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.12.1" />
    <PackageReference Include="Microsoft.IdentityModel.Protocols" Version="6.12.1" />
  </ItemGroup>
```

**Comment by stap123** [comment]
Not sure if it's useful but I can also confirm that we see this issue when upgrading from `Microsoft.IdentityModel.Jwt` version `6.12.0` to version `6.12.2`.

We get the following exception details:

```
dbug: Microsoft.Identity.Web.Resource.JwtBearerMiddlewareDiagnostics[0]
      Begin OnMessageReceivedAsync.
dbug: Microsoft.Identity.Web.Resource.JwtBearerMiddlewareDiagnostics[0]
      End OnMessageReceivedAsync.
fail: Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler[3]
      Exception occurred while processing message.
      System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://login.microsoftonline.com/[TENANT_ID]/v2.0/.well-known/openid-configuration'.
       ---> System.MissingMethodException: Method not found: 'Void Microsoft.IdentityModel.Logging.IdentityModelTelemetryUtil.SetTelemetryData(System.Net.Http.HttpRequestMessage)'.
         at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)
         at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
         at Microsoft.IdentityModel.Protocols.HttpDocumentRetriever.GetDocumentAsync(String address, CancellationToken cancel)
         at Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(String address, IDocumentRetriever retriever, CancellationToken cancel)
         at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
         --- End of inner exception stack trace ---
         at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
         at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
dbug: Microsoft.Identity.Web.Resource.JwtBearerMiddlewareDiagnostics[0]
      Begin OnAuthenticationFailedAsync.
dbug: Microsoft.Identity.Web.Resource.JwtBearerMiddlewareDiagnostics[0]
      End OnAuthenticationFailedAsync.
```

We are using `Microsoft.Identity.Web` version `1.15.2` and have no direct dependency on the `Microsoft.IdentityModel.Protocols` package.

I can confirm that adding a reference to version `6.12.2` of `Microsoft.IdentityModel.Protocols` seems to resolve the issue. Is adding that reference the correct thing to be doing?

Hope that's useful for something.

**Comment by jennyf19** [comment]
@stap123 can you try with Microsoft.Identity.Web version 1.16.0, which references Microsoft.IdentityModel.* as 6.*, so hopefully that works for you without a direct reference.

**Comment by stap123** [comment]
@jennyf19 Yes it does. 

Upgrading to version `1.16.0` of `Microsoft.Identity.Web` and removing the direct reference to `Microsoft.IdentityModel.Protocols` makes everything work as before. 

Thank you 😃

**Comment by brentschmaltz** [comment]
We fixed this with our 6.14.1 release.

**Comment by jfranki** [comment]
This is broken again with version `6.15.0`. I spent the whole day figuring out why clients could not connect to a SignalR hub with JWT Bearer token authentication after updating to said version, while working on `6.14.1`.

In my case, I always had this error in the call to `<hub>/negotiate`, where the previously issued token (through `<hub>/authenticate`) is validated:

```
2021-12-10 16:56:29.2079799 +01:00 [INFORMATION] Failed to validate the token.
System.MissingMethodException: Method not found: 'Void Microsoft.IdentityModel.Tokens.InternalValidators.ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(Microsoft.IdentityModel.Tokens.SecurityToken, System.Nullable`1<System.DateTime>, System.Nullable`1<System.DateTime>, System.String, Microsoft.IdentityModel.Tokens.TokenValidationParameters, System.Text.StringBuilder)'.
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateSignature(String token, TokenValidationParameters validationParameters)
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.ValidateToken(String token, TokenValidationParameters validationParameters, SecurityToken& validatedToken)
   at Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler.HandleAuthenticateAsync()
2021-12-10 16:56:29.2389171 +01:00 [INFORMATION] "Bearer" was not authenticated. Failure message: "Method not found: 'Void Microsoft.IdentityModel.Tokens.InternalValidators.ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(Microsoft.IdentityModel.Tokens.SecurityToken, System.Nullable`1<System.DateTime>, System.Nullable`1<System.DateTime>, System.String, Microsoft.IdentityModel.Tokens.TokenValidationParameters, System.Text.StringBuilder)'."
2021-12-10 16:56:29.2420674 +01:00 [INFORMATION] Authorization failed. "These requirements were not met:\r\nDenyAnonymousAuthorizationRequirement: Requires an authenticated user."
2021-12-10 16:56:29.2465089 +01:00 [INFORMATION] AuthenticationScheme: "Bearer" was challenged.
```
